File: ios/RNJWPlayer/RNJWPlayerAds.m
//
//  RNJWPlayerAds.m
//  RNJWPlayer
//
//  Created by Chaim Paneth on 19/12/2023.
//

#import "RNJWPlayerAds.h"

@implementation RNJWPlayerAds

+ (JWAdvertisingConfig *)configureVASTWithAds:(NSDictionary *)ads error:(JWError **)error {
    JWAdsAdvertisingConfigBuilder* adConfigBuilder = [[JWAdsAdvertisingConfigBuilder alloc] init];
    
    // Configure VAST specific settings here
    // Example: setting ad schedule, tag, etc.
    
    NSArray* scheduleArray = [self getAdSchedule:ads error:error];
    if (scheduleArray.count > 0) {
        [adConfigBuilder schedule:scheduleArray];
    }
    
    id tag = ads[@"tag"];
    if (tag != nil && (tag != (id)[NSNull null])) {
        NSString* encodedString = [tag stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
        NSURL* tagUrl = [NSURL URLWithString:encodedString];
        if (tagUrl != nil) {
            [adConfigBuilder tag:tagUrl];
        }
    }
    
    id adVmap = ads[@"adVmap"];
    if (adVmap != nil && (adVmap != (id)[NSNull null])) {
        NSString* encodedString = [adVmap stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
        NSURL* adVmapUrl = [NSURL URLWithString:encodedString];
        if (adVmapUrl != nil) {
            [adConfigBuilder vmapURL:adVmapUrl];
        }
    }
    
    id openBrowserOnAdClick = ads[@"openBrowserOnAdClick"];
    if (openBrowserOnAdClick != nil && (openBrowserOnAdClick != (id)[NSNull null])) {
        [adConfigBuilder openBrowserOnAdClick:openBrowserOnAdClick];
    }
    
    NSDictionary *adRulesDict = ads[@"adRules"];
    if (adRulesDict) {
        JWAdRules* adRules = [self getAdRules:adRulesDict error:error];
        [adConfigBuilder adRules:adRules];
    }
    
    NSDictionary *adSettingsDict = ads[@"adSettings"];
    if (adSettingsDict) {
        JWAdSettings* adSettings = [self getAdSettings:adSettingsDict];
        [adConfigBuilder adSettings:adSettings];
    }
    
    return [adConfigBuilder buildAndReturnError:error];
}

+ (JWAdvertisingConfig *)configureIMAWithAds:(NSDictionary *)ads error:(JWError **)error {
    JWImaAdvertisingConfigBuilder* adConfigBuilder = [[JWImaAdvertisingConfigBuilder alloc] init];
    
    // Configure Google IMA specific settings here
    // Example: setting ad schedule, tag, IMA settings, etc.
    
    NSArray* scheduleArray = [self getAdSchedule:ads error:error];
    if (scheduleArray.count > 0) {
        [adConfigBuilder schedule:scheduleArray];
    }
    
    id tag = ads[@"tag"];
    if (tag != nil && (tag != (id)[NSNull null])) {
        NSString* encodedString = [tag stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
        NSURL* tagUrl = [NSURL URLWithString:encodedString];
        if (tagUrl != nil) {
            [adConfigBuilder tag:tagUrl];
        }
    }
    
    id adVmap = ads[@"adVmap"];
    if (adVmap != nil && (adVmap != (id)[NSNull null])) {
        NSString* encodedString = [adVmap stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
        NSURL* adVmapUrl = [NSURL URLWithString:encodedString];
        if (adVmapUrl != nil) {
            [adConfigBuilder vmapURL:adVmapUrl];
        }
    }
    
    NSDictionary *adRulesDict = ads[@"adRules"];
    if (adRulesDict) {
        JWAdRules* adRules = [self getAdRules:adRulesDict error:error];
        [adConfigBuilder adRules:adRules];
    }
    
    NSDictionary *imaSettingsDict = ads[@"imaSettings"];
    if (imaSettingsDict) {
        JWImaSettings* imaSettings = [self getIMASettings:imaSettingsDict];
        [adConfigBuilder imaSettings:imaSettings];
    }
    
    return [adConfigBuilder buildAndReturnError:error];
}

+ (JWAdvertisingConfig *)configureIMADAIWithAds:(NSDictionary *)ads error:(JWError **)error {
    JWImaDaiAdvertisingConfigBuilder* adConfigBuilder = [[JWImaDaiAdvertisingConfigBuilder alloc] init];
    
    // Configure Google IMA DAI specific settings here
    // Example: setting stream configuration, friendly obstructions, etc.
    
    NSDictionary *imaSettingsDict = ads[@"imaSettings"];
    if (imaSettingsDict) {
        JWImaSettings* imaSettings = [self getIMASettings:imaSettingsDict];
        [adConfigBuilder imaSettings:imaSettings];
    }
    
    NSDictionary *googleDAIStreamDict = ads[@"googleDAIStream"];
    if (googleDAIStreamDict) {
        JWGoogleDAIStream* googleDAIStream = [self getGoogleDAIStream:googleDAIStreamDict error:error];
        [adConfigBuilder googleDAIStream:googleDAIStream];
    }
    
    return [adConfigBuilder buildAndReturnError:error];
}

+ (NSArray *)getAdSchedule:(NSDictionary *)ads error:(JWError **)error {
    id schedule = ads[@"adSchedule"];
    if(schedule != nil && (schedule != (id)[NSNull null])) {
        NSArray* scheduleAr = (NSArray*)schedule;
        if (scheduleAr.count > 0) {
            NSMutableArray <JWAdBreak*>* scheduleArray = [[NSMutableArray alloc] init];
            
            for (id item in scheduleAr) {
                NSString *offsetString = [item objectForKey:@"offset"];
                NSString *tag = [item objectForKey:@"tag"];
                NSString* encodedString = [tag stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]];
                NSURL* tagUrl = [NSURL URLWithString:encodedString];
                
                if (tagUrl != nil) {
                    JWAdBreakBuilder* adBreakBuilder = [[JWAdBreakBuilder alloc] init];
                    JWAdOffset* offset = [JWAdOffset fromString:offsetString];
                    
                    [adBreakBuilder offset:offset];
                    [adBreakBuilder tags:@[tagUrl]];
                    
                    JWAdBreak *adBreak = [adBreakBuilder buildAndReturnError:error];
                    
                    [scheduleArray addObject:adBreak];
                }
            }
        
            return scheduleArray;
        }
    }
    
    return @[];
}

+ (JWAdRules *)getAdRules:(NSDictionary *)adRulesDict error:(JWError **)error {
    JWAdRulesBuilder *builder = [[JWAdRulesBuilder alloc] init];

    NSUInteger startOn = [adRulesDict[@"startOn"] unsignedIntegerValue];
    NSUInteger frequency = [adRulesDict[@"frequency"] unsignedIntegerValue];
    NSUInteger timeBetweenAds = [adRulesDict[@"timeBetweenAds"] unsignedIntegerValue];
    JWAdShownOnSeek startOnSeek = [self mapStringToJWAdShownOnSeek:adRulesDict[@"startOnSeek"]];

    [builder jwRulesWithStartOn:startOn frequency:frequency timeBetweenAds:timeBetweenAds startOnSeek:startOnSeek];

    return [builder buildAndReturnError:error];
}

+ (JWAdShownOnSeek)mapStringToJWAdShownOnSeek:(NSString *)seekString {
    if ([seekString isEqualToString:@"pre"]) {
        return JWAdShownOnSeekPre;
    }
    return JWAdShownOnSeekNone;
}

+  (JWAdSettings *)getAdSettings:(NSDictionary *)settingsDict {
    JWAdSettingsBuilder *builder = [[JWAdSettingsBuilder alloc] init];
    
    if (settingsDict[@"allowsBackgroundPlayback"]) {
        BOOL allowsBackgroundPlayback = [settingsDict[@"allowsBackgroundPlayback"] boolValue];
        [builder allowsBackgroundPlayback:allowsBackgroundPlayback];
    }

    // Add other settings as needed

    return [builder build];
}

+ (JWImaSettings *)getIMASettings:(NSDictionary *)imaSettingsDict {
    JWImaSettingsBuilder *builder = [[JWImaSettingsBuilder alloc] init];
    if (imaSettingsDict[@"locale"]) {
        [builder locale:imaSettingsDict[@"locale"]];
    }
    if (imaSettingsDict[@"ppid"]) {
        [builder ppid:imaSettingsDict[@"ppid"]];
    }
    if (imaSettingsDict[@"maxRedirects"]) {
        [builder maxRedirects:[imaSettingsDict[@"maxRedirects"] unsignedIntegerValue]];
    }
    if (imaSettingsDict[@"sessionID"]) {
        [builder sessionID:imaSettingsDict[@"sessionID"]];
    }
    if (imaSettingsDict[@"debugMode"]) {
        [builder debugMode:[imaSettingsDict[@"debugMode"] boolValue]];
    }
    return [builder build];
}

+ (JWGoogleDAIStream *)getGoogleDAIStream:(NSDictionary *)googleDAIStreamDict error:(JWError **)error {
    JWGoogleDAIStreamBuilder *builder = [[JWGoogleDAIStreamBuilder alloc] init];
    if (googleDAIStreamDict[@"videoID"] && googleDAIStreamDict[@"cmsID"]) {
        [builder vodStreamInfoWithVideoID:googleDAIStreamDict[@"videoID"] cmsID:googleDAIStreamDict[@"cmsID"]];
    } else if (googleDAIStreamDict[@"assetKey"]) {
        [builder liveStreamInfoWithAssetKey:googleDAIStreamDict[@"assetKey"]];
    }
    if (googleDAIStreamDict[@"apiKey"]) {
        [builder apiKey:googleDAIStreamDict[@"apiKey"]];
    }
    if (googleDAIStreamDict[@"adTagParameters"]) {
        [builder adTagParameters:googleDAIStreamDict[@"adTagParameters"]];
    }
    return [builder buildAndReturnError:error];
}

//+ (UIView *)findViewWithId:(NSString *)viewId {
//    // needs implementation
//}

//+ (NSArray<JWFriendlyObstruction *> *)createFriendlyObstructionsFromArray:(NSArray *)obstructionsArray {
//    NSMutableArray<JWFriendlyObstruction *> *obstructions = [NSMutableArray array];
//
//    for (NSDictionary *obstructionDict in obstructionsArray) {
//        NSString *viewId = obstructionDict[@"viewId"];
//        UIView *view = [self findViewWithId:viewId]; // Implement this method to find the view by ID
//        JWFriendlyObstructionPurpose purpose = [self mapStringToJWFriendlyObstructionPurpose:obstructionDict[@"purpose"]];
//        NSString *reason = obstructionDict[@"reason"];
//
//        JWFriendlyObstruction *obstruction = [[JWFriendlyObstruction alloc] initWithView:view purpose:purpose reason:reason];
//        [obstructions addObject:obstruction];
//    }
//
//    return obstructions;
//}

//+ (JWFriendlyObstructionPurpose)mapStringToJWFriendlyObstructionPurpose:(NSString *)purposeString {
//    if ([purposeString isEqualToString:@"mediaControls"]) {
//        return JWFriendlyObstructionPurposeMediaControls;
//    } else if ([purposeString isEqualToString:@"closeAd"]) {
//        return JWFriendlyObstructionPurposeCloseAd;
//    } else if ([purposeString isEqualToString:@"notVisible"]) {
//        return JWFriendlyObstructionPurposeNotVisible;
//    }
//    return JWFriendlyObstructionPurposeOther;
//}

//+ (JWCompanionAdSlot *)createCompanionAdSlotFromDictionary:(NSDictionary *)companionAdSlotDict {
//    UIView *view = [self findViewWithId:companionAdSlotDict[@"viewId"]]; // Implement this method to find the UIView
//    CGSize size = CGSizeZero;
//    if (companionAdSlotDict[@"size"]) {
//        size = CGSizeMake([companionAdSlotDict[@"size"][@"width"] floatValue],
//                          [companionAdSlotDict[@"size"][@"height"] floatValue]);
//    }
//    JWCompanionAdSlot *slot = [[JWCompanionAdSlot alloc] initWithView:view size:size];
//    return slot;
//}

@end

File: ios/RNJWPlayer/RNJWPlayerAds.swift
//
//  RNJWPlayerAds.swift
//  RNJWPlayer
//
//  Created by Chaim Paneth on 24/12/2023.
//

import Foundation
import JWPlayerKit

class RNJWPlayerAds {

    // Convert configureVASTWithAds function
    static func configureVAST(with ads: [String: Any]) -> JWAdvertisingConfig? {
        let adConfigBuilder = JWAdsAdvertisingConfigBuilder()
        
        // Configure VAST specific settings here
        // Example: setting ad schedule, tag, etc.
        
        if let scheduleArray = getAdSchedule(from: ads), !scheduleArray.isEmpty {
            adConfigBuilder.schedule(scheduleArray)
        }
        
        if let tag = ads["tag"] as? String, let encodedString = tag.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed), let tagUrl = URL(string: encodedString) {
            adConfigBuilder.tag(tagUrl)
        }
        
        if let adVmap = ads["adVmap"] as? String, let encodedString = adVmap.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed), let adVmapUrl = URL(string: encodedString) {
            adConfigBuilder.vmapURL(adVmapUrl)
        }
        
        if let openBrowserOnAdClick = ads["openBrowserOnAdClick"] as? Bool {
            adConfigBuilder.openBrowserOnAdClick(openBrowserOnAdClick)
        }
        
        if let adRulesDict = ads["adRules"] as? [String: Any], let adRules = getAdRules(from: adRulesDict) {
            adConfigBuilder.adRules(adRules)
        }
        
        if let adSettingsDict = ads["adSettings"] as? [String: Any] {
            if let adSettings = getAdSettings(from: adSettingsDict) {
                adConfigBuilder.adSettings(adSettings)
            }
        }
        
        return try? adConfigBuilder.build()
    }

    // Convert configureIMAWithAds function
    static func configureIMA(with ads: [String: Any]) -> JWAdvertisingConfig? {
        // Ensure Google IMA SDK is available
#if !USE_GOOGLE_IMA
    assertionFailure("Error: GoogleAds-IMA-iOS-SDK is not installed. Add $RNJWPlayerUseGoogleIMA = true; to your podfile")
#endif
        
        let adConfigBuilder = JWImaAdvertisingConfigBuilder()
        
        // Configure Google IMA specific settings here
        // Example: setting ad schedule, tag, IMA settings, etc.
        
        if let scheduleArray = getAdSchedule(from: ads), !scheduleArray.isEmpty {
            adConfigBuilder.schedule(scheduleArray)
        }
        
        if let tag = ads["tag"] as? String, let encodedString = tag.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed), let tagUrl = URL(string: encodedString) {
            adConfigBuilder.tag(tagUrl)
        }
        
        if let adVmap = ads["adVmap"] as? String, let encodedString = adVmap.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed), let adVmapUrl = URL(string: encodedString) {
            adConfigBuilder.vmapURL(adVmapUrl)
        }
        
        if let adRulesDict = ads["adRules"] as? [String: Any], let adRules = getAdRules(from: adRulesDict) {
            adConfigBuilder.adRules(adRules)
        }
        
        if let imaSettingsDict = ads["imaSettings"] as? [String: Any] {
            if let imaSettings = getIMASettings(from: imaSettingsDict) {
                adConfigBuilder.imaSettings(imaSettings)
            }
        }
        
        return try? adConfigBuilder.build()
    }
    
    // Convert configureIMADAIWithAds function
    static func configureIMADAI(with ads: [String: Any]) -> JWAdvertisingConfig? {
        // Ensure Google IMA SDK is available
#if !USE_GOOGLE_IMA
    assertionFailure("Error: GoogleAds-IMA-iOS-SDK is not installed. Add $RNJWPlayerUseGoogleIMA = true; to your podfile")
#endif
        
        let adConfigBuilder = JWImaDaiAdvertisingConfigBuilder()
        
        // Configure Google IMA DAI specific settings here
        // Example: setting stream configuration, friendly obstructions, etc.
        
        if let imaSettingsDict = ads["imaSettings"] as? [String: Any] {
            if let imaSettings = getIMASettings(from: imaSettingsDict) {
                adConfigBuilder.imaSettings(imaSettings)
            }
        }
        
        if let googleDAIStreamDict = ads["googleDAIStream"] as? [String: Any], let googleDAIStream = getGoogleDAIStream(from: googleDAIStreamDict) {
            adConfigBuilder.googleDAIStream(googleDAIStream)
        }
        
        return try? adConfigBuilder.build()
    }

    // Convert getAdSchedule function
    static func getAdSchedule(from ads: [String: Any]) -> [JWAdBreak]? {
        guard let schedule = ads["adSchedule"] as? [[String: Any]], !schedule.isEmpty else { return nil }
        
        var scheduleArray: [JWAdBreak] = []
        
        for item in schedule {
            if let offsetString = item["offset"] as? String, let tag = item["tag"] as? String, let encodedString = tag.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed), let tagUrl = URL(string: encodedString) {
                
                let adBreakBuilder = JWAdBreakBuilder()
                if let offset = JWAdOffset.from(string: offsetString) {
                    adBreakBuilder.offset(offset)
                    adBreakBuilder.tags([tagUrl])
                    
                    if let adBreak = try? adBreakBuilder.build() {
                        scheduleArray.append(adBreak)
                    }
                }
            }
        }
        
        return scheduleArray.isEmpty ? nil : scheduleArray
    }

    // Convert getAdRules function
    static func getAdRules(from adRulesDict: [String: Any]) -> JWAdRules? {
        let builder = JWAdRulesBuilder()

        if let startOn = adRulesDict["startOn"] as? UInt, let frequency = adRulesDict["frequency"] as? UInt, let timeBetweenAds = adRulesDict["timeBetweenAds"] as? UInt {
            let startOnSeek = mapStringToJWAdShownOnSeek(adRulesDict["startOnSeek"] as? String)
            builder.jwRules(startOn: startOn, frequency: frequency, timeBetweenAds: timeBetweenAds, startOnSeek: startOnSeek)
        }
        
        return try? builder.build()
    }

    // Convert mapStringToJWAdShownOnSeek function
    static func mapStringToJWAdShownOnSeek(_ seekString: String?) -> JWAdShownOnSeek {
        guard let seekString = seekString else { return .none }
        switch seekString {
        case "pre":
            return .pre
        default:
            return .none
        }
    }

    // Convert getAdSettings function
    static func getAdSettings(from settingsDict: [String: Any]) -> JWAdSettings? {
        let builder = JWAdSettingsBuilder()
        
        if let allowsBackgroundPlayback = settingsDict["allowsBackgroundPlayback"] as? Bool {
            builder.allowsBackgroundPlayback(allowsBackgroundPlayback)
        }
        
        // Add other settings as needed
        
        return builder.build()
    }

    // Convert getIMASettings function
    static func getIMASettings(from imaSettingsDict: [String: Any]) -> JWImaSettings? {
        let builder = JWImaSettingsBuilder()
        if let locale = imaSettingsDict["locale"] as? String {
            builder.locale(locale)
        }
        if let ppid = imaSettingsDict["ppid"] as? String {
            builder.ppid(ppid)
        }
        if let maxRedirects = imaSettingsDict["maxRedirects"] as? UInt {
            builder.maxRedirects(maxRedirects)
        }
        if let sessionID = imaSettingsDict["sessionID"] as? String {
            builder.sessionID(sessionID)
        }
        if let debugMode = imaSettingsDict["debugMode"] as? Bool {
            builder.debugMode(debugMode)
        }
        return builder.build()
    }

    // Convert getGoogleDAIStream function
    static func getGoogleDAIStream(from googleDAIStreamDict: [String: Any]) -> JWGoogleDAIStream? {
        let builder = JWGoogleDAIStreamBuilder()
        if let videoID = googleDAIStreamDict["videoID"] as? String, let cmsID = googleDAIStreamDict["cmsID"] as? String {
            builder.vodStreamInfo(videoID: videoID, cmsID: cmsID)
        } else if let assetKey = googleDAIStreamDict["assetKey"] as? String {
            builder.liveStreamInfo(assetKey: assetKey)
        }
        if let apiKey = googleDAIStreamDict["apiKey"] as? String {
            builder.apiKey(apiKey)
        }
        if let adTagParameters = googleDAIStreamDict["adTagParameters"] as? [String: Any] {
            let stringParams = adTagParameters.compactMapValues { $0 as? String }
            builder.adTagParameters(stringParams)
        }
        return try? builder.build()
    }

    // Placeholder for findViewWithId function - Needs implementation
//    static func findView(withId viewId: String) -> UIView? {
//        // Implementation needed to find and return the view with the given id
//        return nil
//    }
//
//    static func createFriendlyObstructions(fromArray obstructionsArray: [[String: Any]]) -> [JWFriendlyObstruction] {
//        var obstructions: [JWFriendlyObstruction] = []
//
//        for obstructionDict in obstructionsArray {
//            if let viewId = obstructionDict["viewId"] as? String,
//               let view = findView(withId: viewId),
//               let purposeString = obstructionDict["purpose"] as? String,
//               let reason = obstructionDict["reason"] as? String {
//                let purpose = mapStringToJWFriendlyObstructionPurpose(purposeString)
//                let obstruction = JWFriendlyObstruction(view: view, purpose: purpose, reason: reason)
//                obstructions.append(obstruction)
//            }
//        }
//
//        return obstructions
//    }
//
//    static func mapStringToJWFriendlyObstructionPurpose(_ purposeString: String) -> JWFriendlyObstructionPurpose {
//        switch purposeString {
//        case "mediaControls":
//            return .mediaControls
//        case "closeAd":
//            return .closeAd
//        case "notVisible":
//            return .notVisible
//        default:
//            return .other
//        }
//    }
//
//    static func createCompanionAdSlot(fromDictionary companionAdSlotDict: [String: Any]) -> JWCompanionAdSlot? {
//        guard let viewId = companionAdSlotDict["viewId"] as? String,
//              let view = findView(withId: viewId),
//              let sizeDict = companionAdSlotDict["size"] as? [String: Float],
//              let width = sizeDict["width"],
//              let height = sizeDict["height"] else {
//            return nil
//        }
//
//        let size = CGSize(width: CGFloat(width), height: CGFloat(height))
//        let slot = JWCompanionAdSlot(view: view, size: size)
//        return slot
//    }
}

